app/services/messaging/bsnl/api.rb

```diff
@@ -0,0 +1,41 @@
+class Messaging::Bsnl::Api
+  HOST = "bulksms.bsnl.in"
+  PORT = 5010
+
+  def initialize
+    unless credentials.values.all?
+      raise Messaging::Bsnl::Error.new("Missing BSNL credentials")
+    end
+  end
+
+  def get_template_details
+    post("/api/Get_Content_Template_Details")["Content_Template_Ids"]
+  end
+
+  private
+
+  def credentials
+    {
+      header: ENV["BSNL_IHCI_HEADER"],
+      message_type: "SI",
+      entity_id: ENV["BSNL_IHCI_ENTITY_ID"],
+      jwt: Configuration.fetch("bsnl_sms_jwt")
+    }
+  end
+
+  def post(path, body = nil)
+    http = Net::HTTP.new(HOST, PORT)
+    http.use_ssl = true
+    request = Net::HTTP::Post.new(path, "Content-Type" => "application/json")
+    request.body = body.to_json if body
+    request["Authorization"] = "Bearer #{credentials[:jwt]}"
+
+    json_or_string(http.request(request).body)
+  end
+
+  def json_or_string(string)
+    JSON.parse(string)
+  rescue JSON::ParserError
```

Comment:
Is the response not predictable here? Wondering why the rescue is needed.

Reply:
The API returns a JSON for valid responses and an empty string when there are errors. I wanted the API to intelligently cast JSON responses to a ruby hash, hence the method. 

Renamed this to `hsh_or_string` in [8b261e6](https://github.com/simpledotorg/simple-server/pull/3646/commits/8b261e6d3a96bab449ea5ad21789336ccf6fbe18) since it returns a hash, not JSON.